Extractors for value class trees.

paulp opened the pull request:
Salvaged from work six months ago on value class bugs. Seemed like
a shame to lose it (look at the "after" in posterasure.) There may
be some modest overlap with some other tree extractors; I suggest
that tree extractors are not amenable to one-size-fits-all-ness,
and modest overlap should be tolerated, maybe even embraced.

I admit ignorance as to what happend with t6028.check.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1737/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2473/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1740/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2476/

xeno-by commented:
Lgtm

xeno-by commented on src/compiler/scala/tools/nsc/ast/TreeInfo.scala:
```diff
@@ -14,6 +14,65 @@ package ast
+    // arg1.op(arg2) where arg1 and arg2 are the same boxed value class. Returns unboxed values.
+    object BinaryOp {
+      def unapply(t: Tree): Option[(Tree, Symbol, Tree)] = t match {
+        case Apply(sel @ Select(Box(t1, arg1), _), Box(t2, arg2) :: Nil) if isValueClassOp(t1, t2) => Some((arg1, sel.symbol, arg2))
```
Maybe this could be simplified using `BinaryOp` declared above?

retronym commented on src/compiler/scala/tools/nsc/transform/PostErasure.scala:
```diff
@@ -21,51 +22,25 @@ trait PostErasure extends InfoTransform with TypingTransformers {
+        case ValueClass.Underlying(arg)                                   => finish(arg)  //      new C(arg).underlying  ==>  arg
+        case AsInstanceOf(arg, tpe) if arg.tpe <:< tpe                    => finish(arg)  //        arg.asInstanceOf[T]  ==>  arg
+        case ValueClass.BinaryOp(lhs, cmp @ (Object_== | Object_!=), rhs) =>              // new C(arg1) == new C(arg2)  ==>  arg1 == arg2
+          finish(localTyper typed (Apply(Select(lhs, cmp.name) setPos tree.pos, rhs :: Nil) setPos tree.pos))
```
Selection by name is risky here in the face of overloading.

paulp replied:
Huh, why would I have done that. Will fix.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1737/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2473/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1740/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2476/

scala-jenkins commented:
Job pr-checkin-per-commit failed for b37e20:
Failed tests:
- run/t6090.scala
  Took 73 min.
  ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
Job pr-rangepos-per-commit failed for b37e20:
Failed tests:
- run/t6090.scala
  Took 96 min.
  ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
Job pr-checkin-per-commit failed for b37e20c9 [(results)](https://scala-webapps.epfl.ch/jenkins/job/pr-checkin-per-commit/28/):
Failed tests:
- run/t6090.scala
  <br><br>Took 73 min.<br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

scala-jenkins commented:
Job pr-rangepos-per-commit failed for b37e20c9 [(results)](https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos-per-commit/33/):
Failed tests:
- run/t6090.scala
  <br><br>Took 96 min.<br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

JamesIry commented:
ping @paulp

paulp committed:
Extractors for value class trees.

Salvaged from work six months ago on value class bugs. Seemed like
a shame to lose it (look at the "after" in posterasure.) There may
be some modest overlap with some other tree extractors; I suggest
that tree extractors are not amenable to one-size-fits-all-ness,
and modest overlap should be tolerated, maybe even embraced.

I admit ignorance as to what happend with t6028.check.

paulp commented:
Had to hunt down the little flaw which had crept in. Better now.

paulp commented:
I have this working nicely now, and it has moved so far from the original patch I'm just going to reopen it.
